frontend/src/models/propertyDefinitionsModel.ts

```diff
@@ -1,163 +1,223 @@
+    propertyDefinitionStorage: PropertyDefinitionStorage
+): void => {
+    // first time we see this, schedule a fetch
+    if (typeof propertyName === 'string' && !(propertyName in propertyDefinitionStorage)) {
+        window.setTimeout(
+            () => propertyDefinitionsModel.findMounted()?.actions.loadPropertyDefinitions([propertyName]),
+            0
+        )
+    }
+}
+
+export const propertyDefinitionsModel = kea<propertyDefinitionsModelType>([
+    path(['models', 'propertyDefinitionsModel']),
+    actions({
+        // public
+        loadPropertyDefinitions: (propertyKeys: string[]) => ({ propertyKeys }),
+        updatePropertyDefinition: (propertyDefinition: PropertyDefinition) => ({ propertyDefinition }),
+        updatePropertyDefinitions: (propertyDefinitions: PropertyDefinition[] | PropertyDefinitionStorage) => ({
+            propertyDefinitions,
+        }),
+        // internal
+        fetchAllPendingDefinitions: true,
     }),
-    reducers: () => ({
-        propertyStorage: [
-            { results: [], next: null, count: 0 } as PropertyDefinitionStorage,
+    reducers({
+        propertyDefinitionStorage: [
+            { ...localProperties } as PropertyDefinitionStorage,
             {
-                updatePropertyDefinition: (state, { property }) => ({
-                    count: state.count,
-                    results: state.results.map((p) => (property.id === p.id ? property : p)),
-                    next: state.next,
+                updatePropertyDefinition: (state, { propertyDefinition }) => ({
+                    ...state,
+                    [propertyDefinition.name]: propertyDefinition,
+                }),
+                updatePropertyDefinitions: (state, { propertyDefinitions }) => ({
+                    ...state,
+                    ...(Array.isArray(propertyDefinitions)
+                        ? Object.fromEntries(propertyDefinitions.map((p) => [p.name, p]))
+                        : propertyDefinitions),
                 }),
             },
         ],
     }),
-    listeners: ({ actions }) => ({
-        loadPropertyDefinitionsSuccess: ({ propertyStorage }) => {
-            if (propertyStorage.next) {
-                actions.loadPropertyDefinitions()
+    listeners(({ actions, values }) => ({
+        loadPropertyDefinitions: async ({ propertyKeys }) => {
+            const { propertyDefinitionStorage } = values
+
+            let mustLoad = false
+            const pendingStateUpdate: PropertyDefinitionStorage = {}
+            for (const key of propertyKeys) {
+                if (
+                    !(key in propertyDefinitionStorage) ||
+                    propertyDefinitionStorage[key] === PropertyDefinitionState.Error
+                ) {
+                    mustLoad = true
+                    pendingStateUpdate[key] = PropertyDefinitionState.Pending
+                }
             }
+
+            // nothing new to do
+            if (!mustLoad) {
```

Comment:
Looks like this could just check `Object.keys(pendingStateUpdate).length`

Reply:
Makes sense and changed it. I used the bool as a form of premature optimization, as running `Object.keys()` does make a new array every time. Probably not worth the extra lines.